chia/wallet/puzzles/cat.clvm

```diff
@@ -0,0 +1,407 @@
+     (defun-inline create_announcement_condition (prev_coin_id prev_subtotal)
+       (list CREATE_COIN_ANNOUNCEMENT
+             (concat 0xcb (concat prev_coin_id prev_subtotal))
+       )
+      )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+     ;; this function takes a condition and returns an integer indicating
+     ;; the value of all output coins created with CREATE_COIN. If it's not
+     ;; a CREATE_COIN condition, it returns 0.
+
+     (defun-inline output_value_for_condition (condition)
+       (if (= (f condition) CREATE_COIN)
+         (f (r (r condition)))
+         0
+       )
+     )
+
+     ;; this function takes a list of conditions and returns an integer indicating
+     ;; the value of all output coins created with CREATE_COIN
+     (defun output_totals (conditions)
+       (if conditions
+         (+ (output_value_for_condition (f conditions)) (output_totals (r conditions)))
+         0
+       )
+     )
+
+     ;; add three conditions to the list of morphed conditions:
+     ;; ASSERT_MY_COIN_ID for `this_coin_info`
+     ;; CREATE_COIN_ANNOUNCEMENT for my announcement
+     ;; ASSERT_COIN_ANNOUNCEMENT for the next coin's announcement
+     (defun-inline generate_final_output_conditions
+       (
+         prev_subtotal
+         this_subtotal
+         morphed_conditions
+         prev_coin_id
+         this_coin_id
+         next_coin_id
+         cat_mod_struct
+       )
+         (c (create_announcement_condition prev_coin_id prev_subtotal)
+           (c (create_assert_next_announcement_condition this_coin_id this_subtotal next_coin_id cat_mod_struct)
+             morphed_conditions)
+         )
+      )
+
+    (defun-inline coin_info_for_coin_bundle (coin_bundle)
+      (f coin_bundle)
+    )
+
+     ;;;;;;;;;;;;;;;;;;;;;;;;;;; lineage checking
+
+     ;; return true iff parent of `this_coin_info` is provably a cat
+     (defun is_parent_cat (
+       cat_mod_struct
+       parent_id
+       (parent_parent_coin_id parent_inner_puzzle_hash parent_amount)
+     )
+
+       (if (= parent_id
+              (sha256 parent_parent_coin_id
+                      (cat_puzzle_hash cat_mod_struct parent_inner_puzzle_hash)
+                      parent_amount
+              )
+           )
+           1
+           (x)
+       )
+     )
+
+     ;; return true iff the lineage proof is valid
+     ;; lineage_proof is of one of two forms:
+     ;;  (1 . (parent_parent_coin_id parent_inner_puzzle_hash parent_amount))
+     ;;  (0 . some_opaque_proof_passed_to_GENESIS_COIN_CHECKER)
+     ;; so the `f` value determines what kind of proof it is, and the `r` value is the proof
+
+     (defun-inline limitations_program_from_cat_struct (cat_mod_struct)
+       (f (r (r cat_mod_struct)))
+     )
+
+     (defun-inline limitations_program_hash_from_cat_struct (cat_mod_struct)
+       (f (r (r (r cat_mod_struct))))
+     )
+
+    (defun is_bundle_valid ((coin_info . lineage_proof) limitations_program Truths extra_delta inner_conditions)
```

Comment:
This function used to be called `is_lineage_proof_valid`. The comments above still refer to the lineage proof. The fact that it no longer does makes it more challenging to understand what's going on.

Reply:
This should be renamed in the latest commit, but I'm not sure about the comments